experiments/ubuntu/cron/src/cron.8
.\"/* Copyright 1988,1990,1993 by Paul Vixie
.\" * All rights reserved
.\" *
.\" * Distribute freely, except: don't remove my name from the source or
.\" * documentation (don't take credit for my work), mark your changes (don't
.\" * get me blamed for your possible bugs), don't alter or remove this
.\" * notice.  May be sold if buildable source is provided to buyer.  No
.\" * warrantee of any kind, express or implied, is included with this
.\" * software; use at your own risk, responsibility for damages (if any) to
.\" * anyone resulting from the use of this software rests entirely with the
.\" * user.
.\" *
.\" * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
.\" * I'll try to keep a version up to date.  I can be reached as follows:
.\" * Paul Vixie          <paul@vix.com>          uunet!decwrl!vixie!paul
.\" */
.\" 
.\" $Id: cron.8,v 2.2 1993/12/28 08:34:43 vixie Exp $
.\" 
.TH CRON 8 "19 April 2010"
.UC 4
.SH NAME
cron \- daemon to execute scheduled commands (Vixie Cron)
.SH SYNOPSIS
cron
.RB [ -f ]
.RB [ -l ]
.RB [ -L
.IR loglevel ]
.SH DESCRIPTION
.I cron
is started automatically from /etc/init.d on entering multi-user
runlevels.
.SH OPTIONS
.TP 8
.B -f
Stay in foreground mode, don't daemonize.
.TP
.B -l
Enable LSB compliant names for /etc/cron.d files. This setting, however, does
not affect the parsing of files under /etc/cron.hourly, /etc/cron.daily,
/etc/cron.weekly or /etc/cron.monthly.
.TP
.B -L loglevel 
Tell cron what to log about \fBjobs\fR (errors are logged regardless of this
value) as the sum of the following values:
.br
.RS 12
.IP \fB1\fR
will log the start of all cron jobs
.IP \fB2\fR
will log the end of all cron jobs
.IP \fB4\fR
will log all failed jobs (exit status != 0)
.IP \fB8\fR
will log the process number of all cron jobs
.RE
.IP
The default is to log the start of all jobs (1). Logging will be disabled
if \fIlevels\fR is set to zero (0). A value of fifteen (15) will select all
options.
.SH NOTES
.PP
.I cron
searches its spool area (/var/spool/cron/crontabs) for crontab
files (which are named after accounts in
/etc/passwd); crontabs found are loaded into memory.  Note that
crontabs in this directory should not be accessed directly -
the
.I crontab
command should be used to access and update them.

.I cron
also reads /etc/crontab, which is in a slightly different format (see
.IR crontab(5) ).
In Debian, the content of /etc/crontab is predefined
to run programs under /etc/cron.hourly, /etc/cron.daily,
/etc/cron.weekly and /etc/cron.monthly. This configuration is specific to
Debian, see the note under 
.B DEBIAN SPECIFIC
below.
in order 

Additionally, in Debian,
.I cron
reads the files in the /etc/cron.d directory.
.I cron
treats the files in /etc/cron.d as in the same way as the /etc/crontab file (they
follow the special format of that file, i.e. they include the 
.I user
field). However, they are independent of /etc/crontab: they do not, for 
example, inherit environment variable settings from it. This change is
specific to Debian see the note under 
.B DEBIAN SPECIFIC
below.

Like /etc/crontab, the files in the /etc/cron.d directory are
monitored for changes. In general, the system administrator should not use /etc/cron.d/,
but use the standard system crontab /etc/crontab.

/etc/crontab and the files in /etc/cron.d must be owned by root, and must not
be group- or other-writable. In contrast to the spool area, the files
under /etc/cron.d or the files under /etc/cron.hourly, /etc/cron.daily,
/etc/cron.weekly and /etc/cron.monthly may also be symlinks,
provided that both the symlink and the file it points to are owned by root. 
The files under /etc/cron.d do not need to be executable, while the files
under /etc/cron.hourly, /etc/cron.daily,
/etc/cron.weekly and /etc/cron.monthly do, as they are run by 
.I run-parts
(see
.IR run-parts(8) 
for more information).

.I cron
then wakes up every minute, examining all stored crontabs, checking
each command to see if it should be run in the current minute.  When
executing commands, any output is mailed to the owner of the crontab
(or to the user named in the MAILTO environment variable in the
crontab, if such exists).  The children copies of cron running these
processes have their name coerced to uppercase, as will be seen in the
syslog and ps output.
.PP
Additionally,
.I cron
checks each minute to see if its spool directory's modtime (or the modtime
on the 
.IR /etc/crontab
file)
has changed, and if it has,
.I cron
will then examine the modtime on all crontabs files and reload those which have
changed.  Thus
.I cron
need not be restarted whenever a crontab file is modified.  Note that the
.IR crontab (1)
command updates the modtime of the spool directory whenever it changes a
crontab.
.PP
Special considerations exist when the clock is changed by less than 3
hours, for example at the beginning and end of daylight savings
time. If the time has moved forwards, those jobs which would have
run in the time that was skipped will be run soon after the change. 
Conversely, if the time has moved backwards by less than 3 hours,
those jobs that fall into the repeated time will not be re-run.
.PP
Only jobs that run at a particular time (not specified as
@hourly, nor with '*' in the hour or minute specifier) are
affected. Jobs which are specified with wildcards are run based on the
new time immediately.
.PP
Clock changes of more than 3 hours are considered to be corrections to
the clock, and the new time is used immediately.
.PP
.I cron
logs its action to the syslog facility 'cron', and logging may be
controlled using the standard 
.IR syslogd(8) 
facility.
.SH ENVIRONMENT
If configured in
.I /etc/default/cron
in Debian systems, the
.I cron
daemon localisation settings environment can be managed through the use of 
.I /etc/environment
or through the use of
.I /etc/default/locale
with values from the latter overriding values from the former. These
files are read and they will be used to setup the LANG, LC_ALL, and
LC_CTYPE environment variables. These variables are then used to set the
charset of mails, which defaults to 'C'.
.PP
This does
.B NOT
affect the environment of tasks running under cron. For more information
on how to modify the environment of tasks, consult 
.IR crontab(5)
\.
.PP
The daemon will use, if present, the definition from
.I /etc/timezone
for the timezone.
.PP
The environment can be redefined in user's crontab definitions but
.I cron
will only handle tasks in a single timezone.

.SH DEBIAN SPECIFIC
.PP
Debian introduces some changes to 
.I cron
that were not originally available upstream. The most significant
changes introduced are:

.IP \(em
Support for /etc/cron.{hourly,daily,weekly,monthly} via /etc/crontab,
.IP \(em
Support for /etc/cron.d (drop-in dir for package crontabs),
.IP \(em
PAM support,
.IP \(em
SELinux support,
.IP \(em
auditlog support,
.IP \(em
DST and other time-related changes/fixes,
.IP \(em
SGID crontab(1) instead of SUID root,
.IP \(em
Debian-specific file locations and commands,
.IP \(em
Debian-specific configuration (/etc/default/cron),
.IP \(em
numerous other smaller features and fixes.

.PP
Support for /etc/cron.hourly, /etc/cron.daily, /etc/cron.weekly and
/etc/cron.monthly is provided in Debian through the default setting
of the /etc/crontab file (see the system-wide example in 
.IR crontab(5) ).
The default sytem-wide crontab contains four tasks: run every hour, every
day, every week and every month. Each of these tasks will execute 
.B run-parts
providing each one of the directories as an argument. These tasks are disabled if
.B anacron 
is installed (except for the hourly task) to prevent conflicts between
both daemons.

As described above, the files under these directories have to be pass
some sanity checks including the following: be executable, be owned by root,
not be writable by group or other and, if symlinks, point to files owned by
root. Additionally, the file names must conform to the filename requirements
of 
.BR run-parts :
they must be entirely made up of letters, digits and can only contain the
special signs underscores ('_') and hyphens ('-'). Any file that does
not conform to these requirements will not be executed by 
.BR run-parts . 
For example, any file containing dots will be ignored.
This is done to prevent cron from running any of the files
that are left by the Debian package management system when handling files in
/etc/cron.d/ as configuration files (i.e. files ending in .dpkg-dist, .dpkg-orig, 
and .dpkg-new).

This feature can be used by system administrators and packages to include
tasks that will be run at defined intervals. Files created by packages in these
directories should be named after the package that supplies them.

.PP
Support for /etc/cron.d is included in the
.I cron 
daemon itself, which handles this location as the system-wide crontab spool.
This directory can contain any file defining tasks following the format
used in /etc/crontab, i.e. unlike the user cron spool, these files must
provide the username to run the task as in the task definition.

Files in this directory have to be owned by root, do not need to be executable
(they are configuration files, just like /etc/crontab) and 
must conform to the same naming convention as used by 
.IR run-parts(8) :
they
must consist solely of upper- and lower-case letters, digits, underscores,
and hyphens. This means that they 
.B cannot
contain any dots. 
If the 
.B -l
option is specified to 
.I cron
(this option can be setup through /etc/default/cron, see below), then they must
conform to the LSB namespace specification, exactly as in the
.B --lsbsysinit
option in 
.IR run-parts .

The intended purpose
of this feature is to allow packages that require
finer control of their scheduling than the /etc/cron.{hourly,daily,weekly,monthly} 
directories to add a crontab file to /etc/cron.d. Such files
should be named after the package that supplies them.


Also, the default configuration of 
.I cron
is controlled by
.I /etc/default/cron
which is read by the init.d script that launches the
.I cron
daemon. This file determines whether 
.I cron
will read the system's environment variables and makes it possible to add 
additional options to the 
.I cron
program before it is executed, either to configure its logging or to define how
it will treat the files under /etc/cron.d.

.SH "SEE ALSO"
crontab(1), crontab(5), run-parts(8)
.SH AUTHOR
Paul Vixie <paul@vix.com> is the author of 
.I cron
and original creator of this manual page. This page has also been modified for
Debian by Steve Greenland, Javier Fernandez-Sanguino and Christian Kastner.



experiments/ubuntu/cron/src/cron.c
/* Copyright 1988,1990,1993,1994 by Paul Vixie
 * All rights reserved
 *
 * Distribute freely, except: don't remove my name from the source or
 * documentation (don't take credit for my work), mark your changes (don't
 * get me blamed for your possible bugs), don't alter or remove this
 * notice.  May be sold if buildable source is provided to buyer.  No
 * warrantee of any kind, express or implied, is included with this
 * software; use at your own risk, responsibility for damages (if any) to
 * anyone resulting from the use of this software rests entirely with the
 * user.
 *
 * Send bug reports, bug fixes, enhancements, requests, flames, etc., and
 * I'll try to keep a version up to date.  I can be reached as follows:
 * Paul Vixie          <paul@vix.com>          uunet!decwrl!vixie!paul
 */

#if !defined(lint) && !defined(LINT)
static char rcsid[] = "$Id: cron.c,v 2.11 1994/01/15 20:43:43 vixie Exp $";
#endif


#define	MAIN_PROGRAM


#include "cron.h"
#include <signal.h>

#include <sys/types.h>
#include <fcntl.h>

static	void	usage __P((void)),
		run_reboot_jobs __P((cron_db *)),
		find_jobs __P((time_min, cron_db *, int, int)),
		set_time __P((int)),
		cron_sleep __P((time_min)),
#ifdef USE_SIGCHLD
		sigchld_handler __P((int)),
#endif
		sighup_handler __P((int)),
		parse_args __P((int c, char *v[]));


static void
usage() {
#if DEBUGGING
	char **dflags;

	fprintf(stderr, "usage:  %s [-x [", ProgramName);
	for (dflags = DebugFlagNames; *dflags; dflags++)
		fprintf(stderr, "%s%s", *dflags, dflags[1] ? "," : "]");
	fprintf(stderr, "]\n");
#else
	fprintf(stderr, "usage: %s\n", ProgramName);
#endif
	exit(ERROR_EXIT);
}


int
main(argc, argv)
	int	argc;
	char	*argv[];
{


	cron_db	database;
	char *cs;

	ProgramName = argv[0];

#if defined(BSD)
	setlinebuf(stdout);
	setlinebuf(stderr);
#endif

	parse_args(argc, argv);


#ifdef USE_SIGCHLD
	(void) signal(SIGCHLD, sigchld_handler);
#else
	(void) signal(SIGCLD, SIG_IGN);
#endif
	(void) signal(SIGHUP, sighup_handler);

        /* Reopen stdin in case some idiot closed it before starting
           us - it will only be closed, but not having it open here
           screws up other things that will be opened */
        if (fdopen(0,"r") == NULL) {
            (void) open("dev/null", 0);
        }


	acquire_daemonlock(0);
	set_cron_uid();
	set_cron_cwd();

#if defined(POSIX)
	setenv("PATH", _PATH_DEFPATH, 1);
#endif

       /* Get the default locale character set for the mail
        * "Content-Type: ...; charset=" header
        */
       setlocale(LC_ALL,""); /* set locale to system defaults or to
                                that specified by any  LC_* env vars */
       setlocale(LC_COLLATE, "C"); /* Except for collation, since load_database() uses a-z */
       /* Except that "US-ASCII" is preferred to "ANSI_x3.4-1968" in MIME,
        * even though "ANSI_x3.4-1968" is the official charset name. */

       uids_log("Before copy nl_langinfo");

       if ( ( cs = nl_langinfo( CODESET ) ) != 0L && 
               strcmp(cs, "ANSI_x3.4-1968") != 0 )
           strncpy( cron_default_mail_charset, cs, MAX_ENVSTR );
       else
           strcpy( cron_default_mail_charset, "US-ASCII" );

       uids_log("After copy nl_langinfo");

	/* if there are no debug flags turned on, fork as a daemon should.
	 */
# if DEBUGGING
	if (DebugFlags) {
# else
	if (0) {
# endif
		(void) fprintf(stderr, "[%d] cron started\n", getpid());
	} else if (!stay_foreground) {
		switch (fork()) {
		case -1:
			log_it("CRON",getpid(),"DEATH","can't fork");
			exit(0);
			break;
		case 0:
			/* child process */
			log_it("CRON",getpid(),"STARTUP","fork ok");
			(void) setsid();
			freopen("/dev/null", "r", stdin);
			freopen("/dev/null", "w", stdout);
			freopen("/dev/null", "w", stderr);
			break;
		default:
			/* parent process should just die */
			_exit(0);
		}
	}

	acquire_daemonlock(0);
	database.head = NULL;
	database.tail = NULL;
	database.sys_mtime = (time_t) 0;
	database.user_mtime = (time_t) 0;
#ifdef DEBIAN
	database.sysd_mtime = (time_t) 0;
#endif
	load_database(&database);

	set_time(TRUE);
	run_reboot_jobs(&database);
	timeRunning = virtualTime = clockTime;

	/*
	 * too many clocks, not enough time (Al. Einstein)
	 * These clocks are in minutes since the epoch (time()/60).
	 * virtualTime: is the time it *would* be if we woke up
	 * promptly and nobody ever changed the clock. It is
	 * monotonically increasing... unless a timejump happens.
	 * At the top of the loop, all jobs for 'virtualTime' have run.
	 * timeRunning: is the time we last awakened.
	 * clockTime: is the time when set_time was last called.
	 */
	while (TRUE) {
                uids_log("At the top of the loop.");
		time_min timeDiff;
		int wakeupKind;

		/* ... wait for the time (in minutes) to change ... */
		do {
			cron_sleep(timeRunning + 1);
			set_time(FALSE);
		} while (clockTime == timeRunning);
		timeRunning = clockTime;

		check_orphans(&database);
		load_database(&database);

		/*
		 * ... calculate how the current time differs from
		 * our virtual clock. Classify the change into one
		 * of 4 cases
		 */
		timeDiff = timeRunning - virtualTime;

		Debug(DSCH, ("[%d] pulse: %d = %d - %d\n",
            	    getpid(), timeDiff, timeRunning, virtualTime));

                uids_log("timeDiff:");
                uids_debug(timeDiff);

		/* shortcut for the most common case */
		if (timeDiff == 1) {
			virtualTime = timeRunning;
			find_jobs(virtualTime, &database, TRUE, TRUE);
		} else {
			wakeupKind = -1;
			if (timeDiff > -(3*MINUTE_COUNT))
				wakeupKind = 0;
			if (timeDiff > 0)
				wakeupKind = 1;
			if (timeDiff > 5)
				wakeupKind = 2;
			if (timeDiff > (3*MINUTE_COUNT))
				wakeupKind = 3;

                        uids_log("WakeupKind");
                        uids_debug(wakeupKind);

			switch (wakeupKind) {
			case 1:
				/*
				 * case 1: timeDiff is a small positive number
				 * (wokeup late) run jobs for each virtual minute
				 * until caught up.
				 */
				Debug(DSCH, ("[%d], normal case %d minutes to go\n",
				    getpid(), timeRunning - virtualTime))
				do {
					if (job_runqueue())
						sleep(10);
					virtualTime++;
					find_jobs(virtualTime, &database, TRUE, TRUE);
				} while (virtualTime< timeRunning);
				break;

			case 2:
				/*
				 * case 2: timeDiff is a medium-sized positive number,
				 * for example because we went to DST run wildcard
				 * jobs once, then run any fixed-time jobs that would
				 * otherwise be skipped if we use up our minute
				 * (possible, if there are a lot of jobs to run) go
				 * around the loop again so that wildcard jobs have
				 * a chance to run, and we do our housekeeping
				 */
				Debug(DSCH, ("[%d], DST begins %d minutes to go\n",
				    getpid(), timeRunning - virtualTime))
				/* run wildcard jobs for current minute */
				find_jobs(timeRunning, &database, TRUE, FALSE);
	
				/* run fixed-time jobs for each minute missed */ 
				do {
					if (job_runqueue())
						sleep(10);
					virtualTime++;
					find_jobs(virtualTime, &database, FALSE, TRUE);
					set_time(FALSE);
				} while (virtualTime< timeRunning &&
				    clockTime == timeRunning);
				break;
	
			case 0:
				/*
				 * case 3: timeDiff is a small or medium-sized
				 * negative num, eg. because of DST ending just run
				 * the wildcard jobs. The fixed-time jobs probably
				 * have already run, and should not be repeated
				 * virtual time does not change until we are caught up
				 */
				Debug(DSCH, ("[%d], DST ends %d minutes to go\n",
				    getpid(), virtualTime - timeRunning))
				find_jobs(timeRunning, &database, TRUE, FALSE);
				break;
			default:
				/*
				 * other: time has changed a *lot*,
				 * jump virtual time, and run everything
				 */
				Debug(DSCH, ("[%d], clock jumped\n", getpid()))
				virtualTime = timeRunning;
				find_jobs(timeRunning, &database, TRUE, TRUE);
			}
		}
		/* jobs to be run (if any) are loaded. clear the queue */
		job_runqueue();
	}

}


#ifdef DEBIAN
#include <sys/stat.h>
#include <fcntl.h>
#endif

static void
run_reboot_jobs(db)
	cron_db *db;
{
	register user		*u;
	register entry		*e;
        int rbfd;
#ifdef DEBIAN
#define REBOOT_FILE "/var/run/crond.reboot"
	/* Run on actual reboot, rather than cron restart */
	if (access(REBOOT_FILE, F_OK) == 0) {
	        /* File exists, return */
     	        log_it("CRON", getpid(),"INFO",
		       "Skipping @reboot jobs -- not system startup");
	        return;
	}
	/* Create the file */
	if ((rbfd = creat(REBOOT_FILE, S_IRUSR&S_IWUSR)) < 0) {
		/* Bad news, bail out */
	        log_it("CRON",getpid(),"DEATH","Can't create reboot check file");
		exit(0);
	} else {
		close(rbfd);
		log_it("CRON", getpid(),"INFO", "Running @reboot jobs");
	}
      

        Debug(DMISC, ("[%d], Debian running reboot jobs\n",getpid()));
    
#endif
        Debug(DMISC, ("[%d], vixie running reboot jobs\n", getpid()));
	for (u = db->head;  u != NULL;  u = u->next) {
		for (e = u->crontab;  e != NULL;  e = e->next) {
			if (e->flags & WHEN_REBOOT) {
				job_add(e, u);
			}
		}
	}
	(void) job_runqueue();
}


static void
find_jobs(vtime, db, doWild, doNonWild)
	time_min vtime;
	cron_db	*db;
	int doWild;
	int doNonWild;
{
	time_t   virtualSecond  = vtime * SECONDS_PER_MINUTE;
	register struct tm 	*tm = gmtime(&virtualSecond);
	register int		minute, hour, dom, month, dow;
	register user		*u;
	register entry		*e;

	/* make 0-based values out of these so we can use them as indicies
	 */
       
	minute = tm->tm_min -FIRST_MINUTE;
	hour = tm->tm_hour -FIRST_HOUR;
	dom = tm->tm_mday -FIRST_DOM;
	month = tm->tm_mon +1 /* 0..11 -> 1..12 */ -FIRST_MONTH;
	dow = tm->tm_wday -FIRST_DOW;

        uids_log("Current time:");
        uids_debug(tm->tm_min);
        uids_debug(tm->tm_hour);
        uids_debug(tm->tm_mday);
        uids_debug(tm->tm_mon);
        uids_debug(tm->tm_wday);

	Debug(DSCH, ("[%d] tick(%d,%d,%d,%d,%d) %s %s\n",
		getpid(), minute, hour, dom, month, dow,
		doWild?" ":"No wildcard",doNonWild?" ":"Wildcard only"))

	/* the dom/dow situation is odd.  '* * 1,15 * Sun' will run on the
	 * first and fifteenth AND every Sunday;  '* * * * Sun' will run *only*
	 * on Sundays;  '* * 1,15 * *' will run *only* the 1st and 15th.  this
	 * is why we keep 'e->dow_star' and 'e->dom_star'.  yes, it's bizarre.
	 * like many bizarre things, it's the standard.
	 */
	for (u = db->head;  u != NULL;  u = u->next) {
		for (e = u->crontab;  e != NULL;  e = e->next) {
			Debug(DSCH|DEXT, ("user [%s:%d:%d:...] cmd=\"%s\"\n",
			    env_get("LOGNAME", e->envp),
			    e->uid, e->gid, e->cmd))
                        uids_log("Thinking of adding job:"); 
                        uids_log(e->cmd);
                        uids_debug(e);
                        uids_debug(u);

                        uids_log("minute");
                        uids_debug(minute);
                        uids_log("hour");
                        uids_debug(hour);
                        uids_log("month");
                        uids_debug(month);

                        int h = bit_test(e->minute, minute) && bit_test(e->hour, hour) && bit_test(e->month, month);

                        int f = ( ((e->flags & DOM_STAR) || (e->flags & DOW_STAR))
			      ? (bit_test(e->dow,dow) && bit_test(e->dom,dom))
			      : (bit_test(e->dow,dow) || bit_test(e->dom,dom)));
                        uids_log("dow");
                        int g = ((doNonWild && !(e->flags & (MIN_STAR|HR_STAR)))
				    || (doWild && (e->flags & (MIN_STAR|HR_STAR))));
                        uids_log("flags");
                        if (h && f) {
                            if (g) {
                                job_add(e, u);
                            }
                        }

                        /**
			if (bit_test(e->minute, minute) &&
			    bit_test(e->hour, hour) &&
			    bit_test(e->month, month) &&
			    ( ((e->flags & DOM_STAR) || (e->flags & DOW_STAR))
			      ? (bit_test(e->dow,dow) && bit_test(e->dom,dom))
			      : (bit_test(e->dow,dow) || bit_test(e->dom,dom)))) {
                                uids_log("Inside if:");
				if((doNonWild && !(e->flags & (MIN_STAR|HR_STAR)))
				    || (doWild && (e->flags & (MIN_STAR|HR_STAR)))) 
					job_add(e, u);
			} */
                        uids_log("After test!");
		}
	}
}


/*
 * Set StartTime and clockTime to the current time.
 * These are used for computing what time it really is right now.
 * Note that clockTime is a unix wallclock time converted to minutes.
 */
static void
set_time(int initialize)
{
    struct tm tm;
    static int isdst;

    StartTime = time(NULL);

    /* We adjust the time to GMT so we can catch DST changes. */
    tm = *localtime(&StartTime);
    if (initialize || tm.tm_isdst != isdst) {
       isdst = tm.tm_isdst;
       GMToff = get_gmtoff(&StartTime, &tm);
       Debug(DSCH, ("[%d] GMToff=%ld\n",
           getpid(), (long)GMToff))
    }
    clockTime = (StartTime + GMToff) / (time_t)SECONDS_PER_MINUTE;
}

/*
 * try to just hit the next minute
 */
static void
cron_sleep(target)
	time_min target;
{
	time_t t;
	int seconds_to_wait;

	t = time(NULL) + GMToff;

	seconds_to_wait = (int)(target * SECONDS_PER_MINUTE - t) + 1;
	Debug(DSCH, ("[%d] TargetTime=%ld, sec-to-wait=%d\n",
	    getpid(), (long)target*SECONDS_PER_MINUTE, seconds_to_wait))

        if (seconds_to_wait > 0 && seconds_to_wait < 65)
            sleep((unsigned int) seconds_to_wait);
}

#ifdef USE_SIGCHLD
static void
sigchld_handler(x) {
	int save_errno = errno;
	WAIT_T		waiter;
	PID_T		pid;

	for (;;) {
#ifdef POSIX
		pid = waitpid(-1, &waiter, WNOHANG);
#else
		pid = wait3(&waiter, WNOHANG, (struct rusage *)0);
#endif
		switch (pid) {
		case -1:
			Debug(DPROC,
				("[%d] sigchld...no children\n", getpid()))
			errno = save_errno;
			return;
		case 0:
			Debug(DPROC,
				("[%d] sigchld...no dead kids\n", getpid()))
			errno = save_errno;
			return;
		default:
			Debug(DPROC,
				("[%d] sigchld...pid #%d died, stat=%d\n",
				getpid(), pid, WEXITSTATUS(waiter)))
		}
	}
	errno = save_errno;
}
#endif /*USE_SIGCHLD*/


static void
sighup_handler(x) {
	log_close();

	/* we should use sigaction for proper signal blocking as this 
	   has a race, but... */
	signal(SIGHUP, sighup_handler);
}


static void
parse_args(argc, argv)
	int	argc;
	char	*argv[];
{
	int	argch;

	log_level = 1;
	stay_foreground = 0;
        lsbsysinit_mode = 0;

        uids_log("In parse args"); 
        uids_debug(argc); 
        uids_log(argv[0]); 
        uids_log(argv[1]); 

	while (EOF != (argch = getopt(argc, argv, "lfx:L:"))) {
                uids_log("In getopt:");
                uids_debug(argch);

		switch (argch) {
		default:
			usage();
		case 'f':
			stay_foreground = 1;
			break;
		case 'x':
			if (!set_debug_flags(optarg))
				usage();
			break;
                case 'l':
                    lsbsysinit_mode = 1;
                    break;
		case 'L':
		    log_level = atoi(optarg);
		    break;
		}
	}
}
